pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,292 @@
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller"
+	jobutil "k8s.io/kubernetes/pkg/controller/job"
+	"k8s.io/kubernetes/pkg/kubectl/scheme"
+	"k8s.io/kubernetes/pkg/util/metrics"
+)
+
+// TTLAfterFinishedController watches for changes of Jobs API objects. Triggered
+// by Job creation and updates, it enqueues Jobs that have finished and passed
+// the Jobs' `.spec.ttlSecondsAfterFinished` time period to the `jobsToCleanup`
+// queue. The TTLAfterFinishedController has workers who consume the
+// `jobsToCleanup` queue and send requests to the API server to delete the Jobs
+// accordingly.
+// This is implemented outside of Job controller for separation of concerns, and
+// because it will be extended to handle other finishable resource types.
+type TTLAfterFinishedController struct {
+	client   clientset.Interface
+	recorder record.EventRecorder
+
+	// jLister can list/get Jobs from the shared informer's store
+	jLister batchlisters.JobLister
+
+	// jStoreSynced returns true if the Job store has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	jListerSynced cache.InformerSynced
+
+	// Jobs that the controller will attempt to delete.
+	jobsToCleanup workqueue.RateLimitingInterface
+
+	// The clock for tracking time
+	clock clock.Clock
+}
+
+// New creates an instance of TTLAfterFinishedController
+func New(jobInformer batchinformers.JobInformer, client clientset.Interface) *TTLAfterFinishedController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: client.CoreV1().Events("")})
+
+	if client != nil && client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("ttl_after_finished_controller", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	tc := &TTLAfterFinishedController{
+		client:        client,
+		recorder:      eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "ttl-after-finished-controller"}),
+		jobsToCleanup: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ttl_jobs_to_delete"),
+	}
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    tc.addJob,
+		UpdateFunc: tc.updateJob,
+	})
+
+	tc.jLister = jobInformer.Lister()
+	tc.jListerSynced = jobInformer.Informer().HasSynced
+
+	tc.clock = clock.RealClock{}
+
+	return tc
+}
+
+// Run starts the workers to clean up Jobs.
+func (tc *TTLAfterFinishedController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	glog.Infof("Starting TTL after finished controller")
+	defer glog.Infof("Shutting down TTL after finished controller")
+
+	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.jListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(tc.runJobCleanupWorker, time.Second, stopCh)
```

Comment:
Given that we have multiple workers it might be worth using JitterUntil with a small jitter factor. Also wonder if it would be worth turning off sliding. (Only if you turn on Jitter)

Alternately firing off the wait.UntilNoSliding  every 1/workers seconds might be nice, though a bit complex.

Reply:
I'd prefer leaving it as-is. Other controllers with multiple workers also use Until.

Reply:
I agree with leaving this as is. I don't see a use case for the jitter here. The `wait.Until` is here to recover from hard failures and to wait a bit in between starting again.